crates/core/src/transaction.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{Connection, Result};
+use std::ops::Deref;
+
+pub enum TransactionBehavior {
+    Deferred,
+    Immediate,
+    Exclusive,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DropBehavior {
+    Rollback,
+    Commit,
+    Ignore,
+    Panic,
+}
+
+pub struct Transaction {
+    conn: Connection,
+    drop_behavior: DropBehavior,
+}
+
+impl Drop for Transaction {
+    fn drop(&mut self) {
+        if self.conn.is_autocommit() {
+            return;
+        }
+        match self.drop_behavior {
+            DropBehavior::Rollback => {
+                self.do_rollback().unwrap();
+            }
+            DropBehavior::Commit => {
+                self.do_commit().unwrap();
+            }
+            DropBehavior::Ignore => {}
+            DropBehavior::Panic => {
+                panic!("Transaction dropped without being committed or rolled back");
```

Comment:
You need to check if we are panicking already https://doc.rust-lang.org/stable/std/thread/fn.panicking.html before you can panic or else a double panic is really bad.  And same with the unwraps above.